src/main/java/com/google/javascript/gents/CommentLinkingPass.java

```diff
@@ -223,12 +291,14 @@ public final boolean shouldTraverse(NodeTraversal nodeTraversal, Node n, Node pa
       } else if (getLastLineOfCurrentComment() == line - 1) {
         // Comment ends just before code
         linkCommentBufferToNode(n);
-      } else if (!hasRemainingComments()) {
-        // Exhausted all comments, output floating comment before current node
```

Comment:
Instead of being smart in interleaving collecting comments and writing them, maybe down the road we can refactor this as a two step process:
- collect all comments (moving them up and down as we see fit)
- emit all comment from NodeComments mapping.
I think this will avoid such problems (if I understand them correctly)